dali/python/nvidia/dali/pipeline.py

```diff
@@ -990,3 +996,86 @@ def iter_setup(self):
         For example, one can use this function to feed the input
         data from NumPy arrays."""
         pass
+
+
+def _discriminate_args(func, **func_kwargs):
+    """Split args on those applicable to Pipeline constructor and the decorated function."""
+    func_argspec = inspect.getfullargspec(func)
+    ctor_argspec = inspect.getfullargspec(Pipeline.__init__)
+
+    ctor_args = {}
+    fn_args = {}
+
+    for farg in func_kwargs.items():
+        if farg[0] in func_argspec.args or farg[0] in func_argspec.kwonlyargs:
+            fn_args[farg[0]] = farg[1]
+        elif farg[0] in ctor_argspec.args or farg[0] in ctor_argspec.kwonlyargs:
+            ctor_args[farg[0]] = farg[1]
+        else:
+            fn_args[farg[0]] = farg[1]
+
+    return ctor_args, fn_args
+
+
+def pipeline_def(fn=None, **pipeline_kwargs):
+    """
+    A decorator which creates a factory of DALI pipelines.
+    DALI processing graph is defined by the decorated function.
+
+    **Usage**
+
+    First, implement a function that defines a DALI pipeline.
+    Return values of such function will denote output from the
+    DALI pipeline. You can decorate this function with ``@pipeline_def``::
+
+        @nvidia.dali.pipeline.pipeline_def
+        def pipe(flip_vertical, flip_horizontal):
+            ''' Creates a DALI pipeline, which returns flipped and original images '''
+            data, _ = fn.file_reader(file_root=images_dir)
+            img = fn.image_decoder(data, device="mixed")
+            flipped = fn.flip(img, horizontal=flip_horizontal, vertical=flip_vertical)
+            return flipped, img
+
+    Please note, that the decorated function will return
+    **DALI's Pipeline object** (see example below).
+    When creating a Pipeline object using the decorated function,
+    you can pass any number of ``Pipeline.__init__`` arguments as
+    keyword-args to the call::
+
+        my_pipe = pipe(0, batch_size=32, num_threads=1, device_id=0, flip_horizontal=1)
+        my_pipe.build()
```

Comment:
Can you maybe add `flipped, img = my_pipe.run()` and explained that those are the outputs of the pipeline defined by the decorated function?

Reply:
Done